File: legacy/src/panel.ts
/**
 * Interactive control panel TUI for a swarm run.
 *
 * Full-screen alternate-buffer layout (same pattern as opencode's TUI):
 * - Top: live run state (cycle, phase, agents, session IDs, token counts)
 * - Middle: guards & thresholds with current values
 * - Bottom: keybindings
 *
 * Arrow keys navigate fields; enter edits; changes write to .swarm/config.json
 * and take effect on the next cycle (the run reads config each cycle).
 *
 * Uses the SDK to probe live session status + message counts for the selected run.
 */
import fs from "node:fs"
import path from "node:path"
import readline from "node:readline"
import * as Registry from "./registry.ts"
import { connectClient, sessionStatus, sessionMessages } from "./opencode.ts"
import { Style, cutVisible, padVisible } from "./style.ts"
import { enterScreen, leaveScreen, paintScreen, installScreenCleanup } from "./screen.ts"
import { loadProjectConfig, type ProjectConfig } from "./project-config.ts"
import { trace } from "./trace.ts"

type Field = {
  key: string
  label: string
  value: string
  hint: string
  editable: boolean
  kind: "text" | "number" | "toggle"
}

type PanelState = {
  rec: Registry.RunRecord
  fields: Field[]
  selected: number
  editing: boolean
  editBuffer: string
  liveStatus: Record<string, { type: string }>
  agentInfo: Array<{ role: string; sessionID: string; status: string; messages: number }>
}

const GUARD_DEFAULTS: Array<{ key: keyof ProjectConfig; label: string; hint: string; kind: "text" | "toggle" }> = [
  { key: "verify", label: "Verify command", hint: "Shell command after auto-commit (empty = skip)", kind: "text" },
  { key: "singleFlight", label: "Single flight", hint: "Refuse second concurrent run on same project", kind: "toggle" },
  { key: "defaultMerge", label: "Default merge", hint: "Merge worker commits after review unless STOP/HOLD", kind: "toggle" },
  { key: "metrics", label: "Metrics JSONL", hint: "Append cycle facts to metrics.jsonl", kind: "toggle" },
  { key: "redactDumps", label: "Redact dumps", hint: "Redact secrets in session dumps", kind: "toggle" },
]

/** Hardcoded thresholds (shown read-only — these are compile-time constants). */
const HARDCODED: Array<{ label: string; value: string; hint: string }> = [
  { label: "Worker rotate threshold", value: "120 messages (growth since fork)", hint: "WORKER_ROTATE_MSG_THRESHOLD" },
  { label: "System rotate interval", value: "8 cycles", hint: "SYSTEM_ROTATE_CYCLE_INTERVAL" },
  { label: "Digest inject interval", value: "3 minutes", hint: "injectIntervalMs in SystemWatch" },
  { label: "Active watch cooldown", value: "8 minutes", hint: "activeWatchCooldownMs in SystemWatch" },
  { label: "Stall threshold", value: "20 minutes", hint: "stallMs — no bus events before stall" },
  { label: "Max turn retries", value: "3 attempts", hint: "maxAttempts in runTurn" },
  { label: "Ambition ratchet", value: "1 intercept then stop", hint: "doneIntercepted — first DONE gets think-bigger turn" },
  { label: "DONE gate streak", value: ">=2 empty ships + no checklist", hint: "gateDoneSignal" },
  { label: "Health poll interval", value: "45 seconds", hint: "healthTimer" },
  { label: "Bus snapshot interval", value: "20 seconds", hint: "busSnapshotTimer" },
  { label: "Heartbeat interval", value: "30 seconds", hint: "heartbeatTimer" },
]

function loadFields(rec: Registry.RunRecord): Field[] {
  const cfg = loadProjectConfig(rec.project)
  const fields: Field[] = GUARD_DEFAULTS.map((g) => {
    const val = (cfg as any)[g.key]
    return {
      key: g.key,
      label: g.label,
      value: g.kind === "toggle" ? (val ? "on" : "off") : (val || "(empty)"),
      hint: g.hint,
      editable: true,
      kind: g.kind as "text" | "number" | "toggle",
    }
  })
  for (const h of HARDCODED) {
    fields.push({ key: h.label, label: h.label, value: h.value, hint: h.hint, editable: false, kind: "text" })
  }
  return fields
}

function saveField(rec: Registry.RunRecord, key: string, value: string): void {
  const file = path.join(rec.project, ".swarm", "config.json")
  let cfg: Record<string, unknown> = {}
  try {
    if (fs.existsSync(file)) cfg = JSON.parse(fs.readFileSync(file, "utf8"))
  } catch (err) { trace("panel.loadConfig", err) }
  // Map display values back to config types
  const guard = GUARD_DEFAULTS.find((g) => g.key === key)
  if (guard?.kind === "toggle") {
    cfg[key] = value === "on"
  } else if (key === "verify") {
    cfg[key] = value === "(empty)" ? "" : value
  } else {
    cfg[key] = value
  }
  try {
    fs.mkdirSync(path.dirname(file), { recursive: true })
    fs.writeFileSync(file, JSON.stringify(cfg, null, 2) + "\n")
  } catch (err) { trace("panel.saveField", err) }
}

async function probeLive(rec: Registry.RunRecord): Promise<{
  status: Record<string, { type: string }>
  agents: Array<{ role: string; sessionID: string; status: string; messages: number }>
}> {
  if (rec.status !== "running" || !Registry.alive(rec.pid)) {
    return { status: {}, agents: [] }
  }
  try {
    const url = `http://127.0.0.1:${rec.port}`
    const client = connectClient(url, rec.project)
    const status = await sessionStatus(client, rec.project)
    const agents: Array<{ role: string; sessionID: string; status: string; messages: number }> = []
    for (const a of rec.agents ?? []) {
      const st = status[a.sessionID]?.type ?? "idle"
      let messages = 0
      try {
        const msgs = await sessionMessages(client, rec.project, a.sessionID)
        messages = Array.isArray(msgs) ? msgs.length : 0
      } catch (err) { trace("panel.probeLive.messages", err) }
      agents.push({ role: a.role, sessionID: a.sessionID, status: st, messages })
    }
    return { status, agents }
  } catch {
    return { status: {}, agents: [] }
  }
}

function render(state: PanelState): string {
  const w = process.stdout.columns ?? 100
  const h = process.stdout.rows ?? 30
  const rec = state.rec
  const eff = Registry.effectiveStatus(rec)
  const lines: string[] = []

  // Header
  lines.push(Style.highlight(`swarm panel — run ${rec.id}`))
  lines.push("")

  // Run state block
  const stateLines = [
    Style.kv("status:", Style.status(eff)),
    Style.kv("cycle:", Style.cyan(String(rec.cycle))),
    Style.kv("phase:", rec.phase ?? "?"),
    Style.kv("project:", cutVisible(rec.project, w - 14)),
    Style.kv("models:", Style.muted(`s=${rec.models.system} w=${rec.models.worker}`)),
  ]
  for (const a of state.agentInfo) {
    const sid = a.sessionID.slice(0, 16)
    const stColor = a.status === "busy" ? Style.warning(a.status) : Style.muted(a.status)
    stateLines.push(Style.kv(`${a.role}:`, `${stColor}  ses=${sid}…  msgs=${a.messages}`))
  }
  lines.push(...stateLines)
  lines.push("")

  // Config fields
  lines.push(Style.bold("Guards & Thresholds"))
  lines.push(Style.muted("─".repeat(Math.min(w - 2, 60))))
  const editableCount = state.fields.filter((f) => f.editable).length
  for (let i = 0; i < state.fields.length; i++) {
    const f = state.fields[i]
    const marker = f.editable ? (i === state.selected ? Style.cyan("❯") : " ") : Style.muted("·")
    const label = cutVisible(f.label, 28)
    const valStr = f.editable ? (f.kind === "toggle" ? (f.value === "on" ? Style.success("on") : Style.danger("off")) : Style.cyan(f.value)) : Style.muted(f.value)
    const valPad = padVisible(valStr, 30)
    const hintStr = i === state.selected ? Style.muted(`  ${f.hint}`) : ""
    const line = `${marker} ${label} ${valPad}${hintStr}`
    lines.push(cutVisible(line, w - 1))
  }

  lines.push("")
  if (state.editing) {
    lines.push(Style.highlight(`Editing: ${state.editBuffer}_`) + Style.muted("  (enter=save  esc=cancel)"))
  } else {
    lines.push(Style.muted("↑/↓ navigate  ·  enter edit  ·  tab toggle  ·  r refresh  ·  q quit"))
  }

  return lines.slice(0, h - 1).join("\r\n")
}

export async function panel(runId?: string): Promise<void> {
  if (!process.stdin.isTTY) {
    console.log(Style.muted("non-interactive terminal — panel requires a TTY"))
    return
  }

  // Resolve run
  let rec: Registry.RunRecord | undefined
  if (runId) {
    rec = Registry.load(runId)
  } else {
    const active = Registry.list().filter((r) => r.status === "running" && Registry.alive(r.pid))
    if (active.length === 1) {
      rec = active[0]
    } else if (active.length > 1) {
      // Simple inline pick — no need for the full pick() TUI
      console.log(Style.muted("Multiple active runs. Specify: swarm panel <run-id>"))
      for (const r of active) {
        console.log(`  ${Style.bold(r.id)}  cycle ${r.cycle}  ${path.basename(r.project)}`)
      }
      return
    } else {
      // Show most recent run (even if stopped)
      const all = Registry.list()
      if (all.length) rec = all[0]
    }
  }

  if (!rec) {
    console.error(Style.error("no runs found"))
    process.exit(1)
  }

  const fields = loadFields(rec)
  const live = await probeLive(rec)
  const state: PanelState = {
    rec,
    fields,
    selected: 0,
    editing: false,
    editBuffer: "",
    liveStatus: live.status,
    agentInfo: live.agents,
  }

  installScreenCleanup()
  enterScreen()

  let refreshTimer: ReturnType<typeof setInterval> | undefined

  const refresh = async () => {
    const fresh = await probeLive(state.rec)
    state.agentInfo = fresh.agents
    state.liveStatus = fresh.status
    // Reload rec to get latest cycle/phase
    const updated = Registry.load(state.rec.id)
    if (updated) {
      state.rec = updated
      // Reload fields too (config may have changed on disk)
      state.fields = loadFields(updated)
    }
    paintScreen(render(state))
  }

  refreshTimer = setInterval(() => { void refresh() }, 5000)

  const finish = () => {
    if (refreshTimer) clearInterval(refreshTimer)
    try { process.stdin.setRawMode(false) } catch (err) { trace("panel.finish.setRawMode", err) }
    process.stdin.removeListener("keypress", onKey)
    process.stdout.removeListener("resize", onResize)
    process.stdin.pause()
    leaveScreen()
  }

  const onKey = (_: string, key: { name?: string; ctrl?: boolean; shift?: boolean }) => {
    if (state.editing) {
      // Edit mode: type to build buffer, enter to save, esc to cancel
      if (key.name === "return") {
        const f = state.fields[state.selected]
        if (f?.editable) {
          if (f.kind === "toggle") {
            saveField(state.rec, f.key, state.editBuffer === "on" ? "on" : "off")
          } else {
            saveField(state.rec, f.key, state.editBuffer || "(empty)")
          }
          // Update field display
          f.value = f.kind === "toggle" ? (state.editBuffer === "on" ? "on" : "off") : (state.editBuffer || "(empty)")
        }
        state.editing = false
        state.editBuffer = ""
      } else if (key.name === "escape" || (key.ctrl && key.name === "c")) {
        state.editing = false
        state.editBuffer = ""
      } else if (key.name === "backspace") {
        state.editBuffer = state.editBuffer.slice(0, -1)
      } else if (key.name === "tab") {
        // Toggle: cycle on/off
        state.editBuffer = state.editBuffer === "on" ? "off" : "on"
      } else if (key.name && key.name.length === 1 && !key.ctrl) {
        state.editBuffer += key.name
      } else if (key.sequence && key.sequence.length === 1 && !key.ctrl) {
        state.editBuffer += key.sequence
      }
      paintScreen(render(state))
      return
    }

    if (key.name === "up") {
      state.selected = Math.max(0, state.selected - 1)
    } else if (key.name === "down") {
      state.selected = Math.min(state.fields.length - 1, state.selected + 1)
    } else if (key.name === "return") {
      const f = state.fields[state.selected]
      if (f?.editable) {
        state.editing = true
        state.editBuffer = f.kind === "toggle" ? f.value : (f.value === "(empty)" ? "" : f.value)
      }
    } else if (key.name === "tab") {
      // Quick toggle for boolean fields
      const f = state.fields[state.selected]
      if (f?.editable && f.kind === "toggle") {
        const newVal = f.value === "on" ? "off" : "on"
        saveField(state.rec, f.key, newVal)
        f.value = newVal
      }
    } else if (key.name === "r") {
      void refresh()
    } else if (key.name === "escape" || key.name === "q" || (key.ctrl && key.name === "c")) {
      finish()
      return
    } else {
      return
    }
    paintScreen(render(state))
  }

  const onResize = () => paintScreen(render(state))
  process.stdout.on("resize", onResize)
  readline.emitKeypressEvents(process.stdin)
  process.stdin.setRawMode(true)
  process.stdin.resume()
  process.stdin.on("keypress", onKey)

  paintScreen(render(state))
}

File: go-swarm/panel.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

// Panel model for bubbletea
type panelModel struct {
	rec       *RunRecord
	fields    []panelField
	selected  int
	editing   bool
	editBuf   string
	agents    []panelAgentInfo
	width     int
	height    int
	quitting  bool
}

type panelField struct {
	key      string
	label    string
	value    string
	hint     string
	editable bool
	kind     string // "text" | "toggle"
}

type panelAgentInfo struct {
	role      string
	sessionID string
	status    string
	messages  int
}

type tickMsg time.Time

func panelTick() tea.Cmd {
	return tea.Tick(5*time.Second, func(t time.Time) tea.Msg {
		return tickMsg(t)
	})
}

type refreshMsg struct {
	agents []panelAgentInfo
	rec    *RunRecord
	fields []panelField
}

func runPanel(runID string) error {
	if !isTTY() {
		fmt.Fprintln(stdout, muted("non-interactive terminal — panel requires a TTY"))
		return nil
	}

	// Resolve run
	var rec *RunRecord
	if runID != "" {
		rec = regLoad(runID)
	} else {
		active := []RunRecord{}
		for _, r := range regList() {
			if r.Status == "running" && regAlive(r.PID) {
				active = append(active, r)
			}
		}
		if len(active) == 1 {
			rec = &active[0]
		} else if len(active) > 1 {
			fmt.Fprintln(stdout, muted("Multiple active runs. Specify: swarm panel <run-id>"))
			for _, r := range active {
				fmt.Fprintf(stdout, "  %s  cycle %d  %s\n", bold(r.ID), r.Cycle, filepath.Base(r.Project))
			}
			return nil
		} else {
			all := regList()
			if len(all) > 0 {
				rec = &all[0]
			}
		}
	}
	if rec == nil {
		return fmt.Errorf("no runs found")
	}

	fields := loadPanelFields(rec)
	agents := probePanelAgents(rec)

	m := panelModel{
		rec:    rec,
		fields: fields,
		agents: agents,
	}
	p := tea.NewProgram(m, tea.WithAltScreen())
	_, err := p.Run()
	return err
}

func loadPanelFields(rec *RunRecord) []panelField {
	cfg := loadProjectConfig(rec.Project)
	var fields []panelField

	// Editable config fields
	guards := []struct {
		key   string
		label string
		hint  string
		kind  string
		get   func(ResolvedProjectConfig) string
	}{
		{"verify", "Verify command", "Shell command after auto-commit (empty = skip)", "text",
			func(c ResolvedProjectConfig) string { return c.Verify }},
		{"singleFlight", "Single flight", "Refuse second concurrent run on same project", "toggle",
			func(c ResolvedProjectConfig) string { return toggleStr(c.SingleFlight) }},
		{"defaultMerge", "Default merge", "Merge worker commits after review unless STOP/HOLD", "toggle",
			func(c ResolvedProjectConfig) string { return toggleStr(c.DefaultMerge) }},
		{"metrics", "Metrics JSONL", "Append cycle facts to metrics.jsonl", "toggle",
			func(c ResolvedProjectConfig) string { return toggleStr(c.Metrics) }},
		{"redactDumps", "Redact dumps", "Redact secrets in session dumps", "toggle",
			func(c ResolvedProjectConfig) string { return toggleStr(c.RedactDumps) }},
	}
	for _, g := range guards {
		val := g.get(cfg)
		if val == "" {
			val = "(empty)"
		}
		fields = append(fields, panelField{
			key: g.key, label: g.label, value: val, hint: g.hint, editable: true, kind: g.kind,
		})
	}

	// Compile-time thresholds (read-only) — keep in sync with constants.go
	hardcoded := []struct {
		label, value, hint string
	}{
		{"Worker rotate threshold", "120 messages (growth since fork)", "workerRotateMsgThreshold"},
		{"System rotate interval", "8 cycles", "systemRotateCycleInterval"},
		{"Digest inject interval", "3 minutes", "digestInjectInterval (only when pending events)"},
		{"Active watch cooldown", "8 minutes", "activeWatchCooldown — lead turn on alert/STALE"},
		{"Stall threshold", "20 minutes", "stallThreshold — bus quiet while busy (soft re-prompt then rotate)"},
		{"Max turn retries", "3 attempts", "maxTurnAttempts"},
		{"Ambition ratchet", "first DONE only (STOP ends immediately)", "doneIntercepted"},
		{"DONE gate streak", ">=2 empty ships + no checklist", "gateDoneSignal / doneGateEmptyStreak"},
		{"Health poll interval", "45 seconds", "healthCheckInterval"},
		{"Heartbeat interval", "60 seconds", "heartbeatInterval (registry only, no log spam)"},
		{"Multi-worker", "default 1 (N>1 experimental)", "--workers flag"},
	}
	for _, h := range hardcoded {
		fields = append(fields, panelField{
			key: h.label, label: h.label, value: h.value, hint: h.hint, editable: false, kind: "text",
		})
	}
	return fields
}

func toggleStr(b bool) string {
	if b {
		return "on"
	}
	return "off"
}

func savePanelField(rec *RunRecord, key, value string) {
	file := filepath.Join(rec.Project, ".swarm", "config.json")
	cfg := map[string]any{}
	if data, err := os.ReadFile(file); err == nil {
		json.Unmarshal(data, &cfg)
	}
	// Map display to config
	switch key {
	case "singleFlight", "defaultMerge", "metrics", "redactDumps":
		cfg[key] = value == "on"
	case "verify":
		if value == "(empty)" {
			cfg[key] = ""
		} else {
			cfg[key] = value
		}
	default:
		cfg[key] = value
	}
	os.MkdirAll(filepath.Dir(file), 0755)
	data, _ := json.MarshalIndent(cfg, "", "  ")
	os.WriteFile(file, data, 0644)
}

func probePanelAgents(rec *RunRecord) []panelAgentInfo {
	if rec.Status != "running" || !regAlive(rec.PID) {
		return nil
	}
	sdk := newSDKClient(fmt.Sprintf("http://127.0.0.1:%d", rec.Port), rec.Project)
	statuses, err := sdk.sessionStatus()
	if err != nil {
		return nil
	}
	var agents []panelAgentInfo
	for _, a := range rec.Agents {
		st := "idle"
		if s, ok := statuses[a.SessionID]; ok {
			st = s.Type
		}
		msgs := 0
		if m, err := sdk.sessionMessages(a.SessionID, 0); err == nil {
			msgs = len(m)
		}
		agents = append(agents, panelAgentInfo{
			role: a.Role, sessionID: a.SessionID, status: st, messages: msgs,
		})
	}
	return agents
}

// bubbletea Model implementation

func (m panelModel) Init() tea.Cmd {
	return panelTick()
}

func (m panelModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.WindowSizeMsg:
		m.width = msg.Width
		m.height = msg.Height
		return m, nil

	case tickMsg:
		// Refresh live state
		if rec := regLoad(m.rec.ID); rec != nil {
			m.rec = rec
			m.fields = loadPanelFields(rec)
		}
		m.agents = probePanelAgents(m.rec)
		return m, panelTick()

	case tea.KeyMsg:
		if m.editing {
			switch msg.String() {
			case "enter":
				f := &m.fields[m.selected]
				if f.editable {
					if f.kind == "toggle" {
						savePanelField(m.rec, f.key, m.editBuf)
					} else {
						savePanelField(m.rec, f.key, m.editBuf)
					}
					f.value = m.editBuf
				}
				m.editing = false
				m.editBuf = ""
			case "esc":
				m.editing = false
				m.editBuf = ""
			case "tab":
				if m.fields[m.selected].kind == "toggle" {
					m.editBuf = toggleStr(m.editBuf == "on")
				}
			case "backspace":
				if len(m.editBuf) > 0 {
					m.editBuf = m.editBuf[:len(m.editBuf)-1]
				}
			default:
				if len(msg.String()) == 1 {
					m.editBuf += msg.String()
				}
			}
			return m, nil
		}

		switch msg.String() {
		case "up", "k":
			if m.selected > 0 {
				m.selected--
			}
		case "down", "j":
			if m.selected < len(m.fields)-1 {
				m.selected++
			}
		case "enter":
			f := m.fields[m.selected]
			if f.editable {
				m.editing = true
				if f.kind == "toggle" {
					m.editBuf = f.value
				} else {
					m.editBuf = f.value
					if m.editBuf == "(empty)" {
						m.editBuf = ""
					}
				}
			}
		case "tab":
			f := &m.fields[m.selected]
			if f.editable && f.kind == "toggle" {
				newVal := "off"
				if f.value == "off" {
					newVal = "on"
				}
				savePanelField(m.rec, f.key, newVal)
				f.value = newVal
			}
		case "r":
			if rec := regLoad(m.rec.ID); rec != nil {
				m.rec = rec
				m.fields = loadPanelFields(rec)
			}
			m.agents = probePanelAgents(m.rec)
		case "q", "esc", "ctrl+c":
			m.quitting = true
			return m, tea.Quit
		}
	}
	return m, nil
}

func (m panelModel) View() string {
	if m.quitting {
		return ""
	}

	var styles = struct {
		title    lipgloss.Style
		label    lipgloss.Style
		value    lipgloss.Style
		hint     lipgloss.Style
		selected lipgloss.Style
		readonly lipgloss.Style
		header   lipgloss.Style
		footer   lipgloss.Style
	}{
		title:    lipgloss.NewStyle().Bold(true).Foreground(lipgloss.Color("36")),
		label:    lipgloss.NewStyle().Foreground(lipgloss.Color("90")),
		value:    lipgloss.NewStyle().Foreground(lipgloss.Color("36")),
		hint:     lipgloss.NewStyle().Foreground(lipgloss.Color("90")),
		selected: lipgloss.NewStyle().Foreground(lipgloss.Color("36")).Bold(true),
		readonly: lipgloss.NewStyle().Foreground(lipgloss.Color("90")),
		header:   lipgloss.NewStyle().Bold(true),
		footer:   lipgloss.NewStyle().Foreground(lipgloss.Color("90")),
	}

	var b strings.Builder

	// Title
	b.WriteString(styles.title.Render("swarm panel — run "+m.rec.ID))
	b.WriteString("\n\n")

	// Run state
	eff := regEffectiveStatus(m.rec)
	b.WriteString(fmt.Sprintf("status: %s  cycle: %s  phase: %s\n",
		statusBadge(eff), cyan(fmt.Sprintf("%d", m.rec.Cycle)), m.rec.Phase))
	b.WriteString(fmt.Sprintf("project: %s\n", truncate(m.rec.Project, 60)))
	b.WriteString(fmt.Sprintf("models: %s\n", muted(fmt.Sprintf("s=%s w=%s", m.rec.Models.System, m.rec.Models.Worker))))

	// Agent info
	for _, a := range m.agents {
		st := a.status
		if st == "busy" {
			st = warning(st)
		}
		b.WriteString(fmt.Sprintf("%s: %s  ses=%s…  msgs=%d\n",
			a.role, st, truncate(a.sessionID, 16), a.messages))
	}
	b.WriteString("\n")

	// Guards & thresholds
	b.WriteString(styles.header.Render("Guards & Thresholds"))
	b.WriteString("\n")
	b.WriteString(muted(strings.Repeat("─", 50)))
	b.WriteString("\n")

	for i, f := range m.fields {
		marker := " "
		if f.editable {
			if i == m.selected {
				marker = styles.selected.Render("❯")
			} else {
				marker = " "
			}
		} else {
			marker = styles.readonly.Render("·")
		}

		labelStr := padRight(f.label, 28)
		valStr := f.value
		if f.editable && f.kind == "toggle" {
			if f.value == "on" {
				valStr = success("on")
			} else {
				valStr = danger("off")
			}
		} else if f.editable {
			valStr = cyan(f.value)
		} else {
			valStr = muted(f.value)
		}
		valPad := padRight(valStr, 30)

		hintStr := ""
		if i == m.selected {
			hintStr = muted("  " + f.hint)
		}

		b.WriteString(fmt.Sprintf("%s %s %s%s\n", marker, labelStr, valPad, hintStr))
	}

	b.WriteString("\n")
	if m.editing {
		b.WriteString(highlight("Editing: " + m.editBuf + "_"))
		b.WriteString(muted("  (enter=save  esc=cancel)"))
	} else {
		b.WriteString(muted("↑/↓ navigate  ·  enter edit  ·  tab toggle  ·  r refresh  ·  q quit"))
	}

	return b.String()
}